### ISSUE TITLE
Testsuite failures with HDF5 enabled.

### ISSUE THREAD

**Issue by ghisvail** [Shark-ML/Shark#43]
I have started rolling out some Debian package builds for `3.0.1`, but the HDF5 tests fail with the following error:
```
        Start 124: Data_HDF5
124/167 Test #124: Data_HDF5 ....................................***Failed    0.07 sec
Running 4 test cases...
unknown location(0): fatal error in "BasicTests": std::exception: [loadIntoMatrix] open file name: ./Test/test_data/testfile_for_import.h5 (FAILED)
unknown location(0): fatal error in "CscTests": std::exception: [loadIntoMatrix] open file name: ./Test/test_data/testfile_for_import.h5 (FAILED)
unknown location(0): fatal error in "OneDimension": std::exception: [loadIntoMatrix] open file name: ./Test/test_data/testfile_for_import.h5 (FAILED)
expected: [importHDF5] Get data set(data/dummy) info from file(./Test/test_data/testfile_for_import.h5). got: [loadIntoMatrix] open file name: ./Test/test_data/testfile_for_import.h5 (FAILED)
/<<BUILDDIR>>/shark-3.0.1+ds1/Test/Data/HDF5Tests.cpp(264): error in "NegativeTests": incorrect exception shark::Exception is caught
expected: [loadIntoMatrix][./Test/test_data/testfile_for_import.h5][data/three_dimension] Support 1 or 2 dimensions, but this dataset has at least 3 dimensions. got: [loadIntoMatrix] open file name: ./Test/test_data/testfile_for_import.h5 (FAILED)
/<<BUILDDIR>>/shark-3.0.1+ds1/Test/Data/HDF5Tests.cpp(279): error in "NegativeTests": incorrect exception shark::Exception is caught
expected: [loadIntoMatrix] DataType doesn't match. HDF5 data type in dataset(./Test/test_data/testfile_for_import.h5::data/data1): 1, size: 8 got: [loadIntoMatrix] open file name: ./Test/test_data/testfile_for_import.h5 (FAILED)
/<<BUILDDIR>>/shark-3.0.1+ds1/Test/Data/HDF5Tests.cpp(294): error in "NegativeTests": incorrect exception shark::Exception is caught

*** 6 failures detected in test suite "CoreHDF5TestModule"
```
resulting in an FTBFS. I could provide a package build without HDF5 enabled, whilst this issue is investigated. The complete build log is available [here](http://debomatic-amd64.debian.net/distribution#unstable/shark/3.0.1+ds1-1/buildlog).

Cheers,
Ghis

**Comment by Ulfgard** [comment]
The errors state that the tests can not find their test files. Unfortunately you are the first one reporting such an issue and everything works for me. we copy the files to $BuildDir/Test/ and assume tests are in $BuildDir

**Comment by ghisvail** [comment]
```
if(HDF5_FOUND)
	file( COPY test_data/ DESTINATION test_data/ )
endif()
```
Is that really doing what you think it does, i.e. copying the test data to their corresponding location under the current binary directory?

**Comment by ghisvail** [comment]
Here is what happens on my machine:
```
mkdir build && cd build
cmake -DBUILD_TESTING=ON ..
CTEST_OUTPUT_ON_FAILURE=1 ctest -R "HDF5"
```
Result:
```
Test project /home/ghislain/debian/packages/shark/build
    Start 124: Data_HDF5
1/1 Test #124: Data_HDF5 ........................***Failed    0.01 sec
Running 4 test cases...
unknown location(0): fatal error in "BasicTests": std::exception: [loadIntoMatrix] open file name: ./Test/test_data/testfile_for_import.h5 (FAILED)
unknown location(0): fatal error in "CscTests": std::exception: [loadIntoMatrix] open file name: ./Test/test_data/testfile_for_import.h5 (FAILED)
unknown location(0): fatal error in "OneDimension": std::exception: [loadIntoMatrix] open file name: ./Test/test_data/testfile_for_import.h5 (FAILED)
expected: [importHDF5] Get data set(data/dummy) info from file(./Test/test_data/testfile_for_import.h5). got: [loadIntoMatrix] open file name: ./Test/test_data/testfile_for_import.h5 (FAILED)
/home/ghislain/debian/packages/shark/Test/Data/HDF5Tests.cpp(264): error in "NegativeTests": incorrect exception shark::Exception is caught
expected: [loadIntoMatrix][./Test/test_data/testfile_for_import.h5][data/three_dimension] Support 1 or 2 dimensions, but this dataset has at least 3 dimensions. got: [loadIntoMatrix] open file name: ./Test/test_data/testfile_for_import.h5 (FAILED)
/home/ghislain/debian/packages/shark/Test/Data/HDF5Tests.cpp(279): error in "NegativeTests": incorrect exception shark::Exception is caught
expected: [loadIntoMatrix] DataType doesn't match. HDF5 data type in dataset(./Test/test_data/testfile_for_import.h5::data/data1): 1, size: 8 got: [loadIntoMatrix] open file name: ./Test/test_data/testfile_for_import.h5 (FAILED)
/home/ghislain/debian/packages/shark/Test/Data/HDF5Tests.cpp(294): error in "NegativeTests": incorrect exception shark::Exception is caught

*** 6 failures detected in test suite "CoreHDF5TestModule"


0% tests passed, 1 tests failed out of 1

Total Test time (real) =   0.01 sec

The following tests FAILED:
	124 - Data_HDF5 (Failed)
Errors while running CTest
```
Additional information:
```
find build/ -name "Data_HDF5"
build/bin/Data_HDF5
find build/ -name "*.h5"
build/Test/test_data/testfile_for_import.h5
```
So launching `Data_HDF5` under its location is unlikely to find the h5 test file. At least that's what CTest reports with the current design.

**Comment by ghisvail** [comment]
I should add that manually calling './bin/Data_HDF5' seems to be working.

**Comment by Ulfgard** [comment]
starting from the top:

1. Yes it does copy it in the right location: https://cmake.org/cmake/help/v3.0/command/file.html

> The COPY signature copies files, directories, and symlinks to a destination folder. Relative input paths are evaluated with respect to the current source directory, and a relative destination is evaluated with respect to the current build directory.

and the file location is correct. This runs fine on my machine...

2. We usually just call "make test" in /build/ which works for us. I have no Idea why this should not be the case for you. I also do not know in which directory Data_HDF5 is actually executed in your case...

**Comment by Ulfgard** [comment]
strange it does now also fail for me. I am on it.

**Comment by Ulfgard** [comment]
should be fixed in master by  2adf799.

thanks! Sorry that everything takes so long...
